### ISSUE TITLE
A lot of macros functions in code

### ISSUE THREAD

**Issue by d9j** [oatpp/oatpp#519]
Hello.  i just noticed the code contains files with a lot of macros functions. for example file ApiClient_define , 
Wouldnt it better to have utility class to handle it with static methods? because for example 'it's not easy to debug these methods /

I would love to help . but i'm not an experienced c++ dev.  :(

**Comment by sgallou** [comment]
I share your opinion. These macros can be useful to quickly develop with oatpp, but it also makes integration in application difficult and creates strong coupling between application logic and oatpp.
On my side, I want my webserver to be abstracted from my app (My app just uses a webserver, without knowing that oatpp is the webserver), I'm using oatpp without macros, and it's sometimes difficult to find the objects to use to create an end-point, use websocket, add swagger doc, ...
In my opinion, reducing the use of macros will be a good evolution for oatpp.

**Comment by Keidan** [comment]
I agree with you, even if I can understand why there are so many macros in oat++, I still find it a shame to see macros in c++ code (11+).
Especially during debugging sessions, macros don't really make it easier to understand...

Of course this doesn't take away from the fact that oat++ remains a clean and well designed code :blush:

**Comment by lganzzzo** [comment]
Hello guys,

I'm open to propositions, and I agree that macros make it difficult to debug.
Also, I hope that miracle happens one day, and C++ implements something like [CircleLang](https://www.circle-lang.org/), and again, hopefully in a clean and `READABLE` way and not like it used to - `___i_have___no_idea__what__this_is___`.

However, I don't know how to implement "sugar" like the following without macros:

```cpp
ENDPOINT("GET", "users/{userId}", getUserById,
         PATH(Int32, userId))
{
  return createDtoResponse(Status::CODE_200, m_userService.getUserById(userId));
}
```

This API is used by a lot of people, and it improves coding time significantly.

---

> These macros can be useful to quickly develop with oatpp, but it also makes integration in application difficult and creates strong coupling between application logic and oatpp.
On my side, I want my webserver to be abstracted from my app (My app just uses a webserver, without knowing that oatpp is the webserver)

Hey @sgallou , can please provide an example when why it's hard to integrate with oatpp when using macros? 


> I agree with you, even if I can understand why there are so many macros in oat++, I still find it a shame to see macros in c++ code (11+).

@Keidan , C++11 is nice and cool compared to 98. 
However, its compile-time metaprogramming abilities are poor as hell.
And when it comes to more-or-less complex things, I would argue what's better - `macros` or `constexpr` with `templates`, and what will produce more readable and less verbose end-user API.

**Comment by sgallou** [comment]
@lganzzzo In my case, I starts integrate oatpp in a application already abstracting it's webserver, my goal was to replace existing implementation by oatpp. But I don't want to break this abstraction layer and couple application logic with oatpp. My coding rules also prohibit macros as much as possible.
To achieve that, I often need to read macro code to understand what object I need and how to use it.
Also, I already have a dependency injection framework and don't want to use OATPP_COMPONENT macros (the same for provided DTO support).

So actually it works (even if development is not finished), and I got a satisfying result.

So in my mind it's more a lack in documentation (a "coding without macros" documentation would be helpful) than a coding or architecture issue. A such documentation might save some hours to me.

Note also than I appreciate the trend (since 1.3.0) to use more standard types (like std::string).

**Comment by lganzzzo** [comment]
Hey @sgallou ,

Thanks for the description and for the note.
Yes, documentation should be improved on this matter.

Also, please feel free to create issues or to post in Gitter, if you have questions regarding macro-free implementation.

**Comment by Keidan** [comment]
Staying in C++11, I think it would be hell to have something clean that would be macro-free, and cover all your needs, especially since the constexpr in C++11 is quite limited.
Also, if I'm not mistaken, it seems to me that you mentioned that you don't want to move to C++17 and above (which I can understand), so it would be a pain to remove macros and preserve readable code.

It's true that a solution like Circle would be really nice and would be very close to what exists in other languages.

I agree with @sgallou that with a documentation/note on how to use oat++ without macros, I would've avoided a few cold sweats when unpacking the macros to get the exact type of the different components 😄

**Comment by d9j** [comment]
i'm grateful for this project. at least something i can use in real world project and push myself to get more experience in C++ along with that.
